File: src/leetcode/java/AverageOfLevelsInBinaryTree_637.java
package leetcode.java;

import java.util.LinkedList;
import java.util.List;

public class AverageOfLevelsInBinaryTree_637 {
  public List<Double> averageOfLevels(TreeNode root) {
    LinkedList<Double> res = new LinkedList<>();
    if (root == null) return res;

    LinkedList<TreeNode> queue = new LinkedList<>();
    queue.addLast(root);
    int levelSize = 0;
    while ((levelSize = queue.size()) > 0) {
      double avg = 0.0;
      int size = 0;
      for (; levelSize > 0; levelSize--) {
        TreeNode node = queue.removeFirst();
        size++;
        avg += (node.val - avg) / size;
        if (node.left != null) queue.addLast(node.left);
        if (node.right != null) queue.addLast(node.right);
      }
      res.add(avg);
    }
    return res;
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
      val = x;
    }
  }
}

File: src/leetcode/python/AverageOfLevelsInBinaryTree_637.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class AverageOfLevelsInBinaryTree_637(object):
    def averageOfLevels(self, root):
        """
        :type root: TreeNode
        :rtype: List[float]
        """
        res = []
        if not root:
            return res
        queue = [root]
        while queue:
            res.append((sum(float(node.val) for node in queue) / len(queue)))
            queue = [child for node in queue for child in (node.left, node.right) if child]
        return res
